### PR TITLE
feat(expr): build expression from pretty string

### LINKED ISSUE #7881: test facility: build `BoxedExpression` from "pretty" representation of expression
### Is your feature request related to a problem? Please describe.

We now have plenty of test cases and benchmark cases building `BoxedExpression` with `build_from_prost` and hand-constructed `ExprNode`. The lack of function like `build_expression_from_pretty("$0 > 10")` makes writing unit tests painful.

An example:

```rust
    // in src/batch/benches/filter.rs

    // Expression: $1 % 2 == 0
    let expr = {
        let input_ref1 = ExprNode {
            expr_type: InputRef as i32,
            return_type: Some(risingwave_pb::data::DataType {
                type_name: TypeName::Int64 as i32,
                ..Default::default()
            }),
            rex_node: Some(RexNode::InputRef(InputRefExpr { column_idx: 0 })),
        };

        let literal2 = ExprNode {
            expr_type: TConstValue as i32,
            return_type: Some(risingwave_pb::data::DataType {
                type_name: TypeName::Int64 as i32,
                ..Default::default()
            }),
            rex_node: Some(RexNode::Constant(ProstDatum {
                body: serialize_datum(Some(ScalarImpl::Int64(2)).as_ref()),
            })),
        };

        // $1 % 2
        let mod2 = ExprNode {
            expr_type: Modulus as i32,
            return_type: Some(risingwave_pb::data::DataType {
                type_name: TypeName::Int64 as i32,
                ..Default::default()
            }),
            rex_node: Some(RexNode::FuncCall(FunctionCall {
                children: vec![input_ref1, literal2],
            })),
        };

        let literal0 = ExprNode {
            expr_type: TConstValue as i32,
            return_type: Some(risingwave_pb::data::DataType {
                type_name: TypeName::Int64 as i32,
                ..Default::default()
            }),
            rex_node: Some(RexNode::Constant(ProstDatum {
                body: serialize_datum(Some(ScalarImpl::Int64(0)).as_ref()),
            })),
        };

        // $1 % 2 == 0
        ExprNode {
            expr_type: Equal as i32,
            return_type: Some(risingwave_pb::data::DataType {
                type_name: TypeName::Boolean as i32,
                ..Default::default()
            }),
            rex_node: Some(RexNode::FuncCall(FunctionCall {
                children: vec![mod2, literal0],
            })),
        }
    };

    let expr = build_from_prost(&expr).unwrap();
```

If we have `build_expression_from_pretty`, this piece of code can be reduced to:

```rust
    let expr = build_expression_from_pretty("$1 % 2 == 0");
```

### Describe the solution you'd like

I think a subset of SQL expression parser is enough, only allowing variables in the form of `$n` representing `InputRef` and basic operators.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR introduces a simple DSL for expressions, so that we can build expressions painless in unit tests. (resolve #7881)

The language is defined as follows:
```
inputref := #<index>:<type>
constant := <value>:<type>
function := (<name>:<type> <child1> <child2>...)
```

For example:
```
#0 % 2 == 0
(equal:boolean (modulus:int8 #0:int8 2:int8) 0:int8)
```

The overall design follows the [egg](https://docs.rs/egg/latest/egg/index.html#simple-example). Except that we have to explicitly specify the return type of each node. It might not be the best and is subject to be changed if anyone has a better idea.

This PR contains a small lexer and parser for the DSL (mainly generated by GPT), and replaces boilerplate code with the new language whenever possible.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] ~~I have added necessary unit tests and integration tests~~
- [x] ~~I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- [x] ~~I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### CODE REVIEW

**Review by richardchien — COMMENTED** [review]
I would suggest `$idx` and `int32`/`int64` instead of `#idx` and `int4`/`int8`, for better consistency with our current system. BTW, it's great to see this work!🚀 Guess we may need some documents for the new DSL.

File: src/common/src/array/data_chunk.rs [hunk 1]
```diff
@@ -579,6 +588,11 @@ impl DataChunkTestExt for DataChunk {
                         return None;
                     }
                     Some(match builder {
+                        ArrayBuilderImpl::Bool(_) => ScalarImpl::Bool(match s {
```
Comment (BugenZhao) [inline src/common/src/array/data_chunk.rs:591]: Can we remove the default branch of this `match`?

**Review by BugenZhao — APPROVED** [review]
LGTM!! 🚀🚀🚀

**Comment by wangrunji0408** [comment]
> I would suggest `$idx` and `int32`/`int64` instead of `#idx` and `int4`/`int8`, for better consistency with our current system. 

When I searched `:int4` in our codebase, I found a lot of occurrence in SQL tests, e.g. `select 1::int4`.
Should we also be consistent with them? 🥲

**Comment by BugenZhao** [comment]
+1 on `int2` / `int4`. 🥵 It's better to follow the terminology of Postgres.

**Comment by richardchien** [comment]
> But when I searched `:int4` in our codebase, I found a lot of occurrence in SQL tests, e.g. `select 1::int4`.
> Should we also be consistent with them? 🥲

My fault 🥵, `int2`/`int4` seems OK

**Review by richardchien — APPROVED** [review]
LGTM!

File: src/common/src/array/data_chunk.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/common/src/array/data_chunk.rs]: You inspired me that we could reuse `ScalarImpl::from_text` for this (except for `f`/`t` ). 😇

**Pull request merged**